shared/domains/chelonia/internals.js

```diff
@@ -169,14 +169,71 @@ export default (sbp('sbp/selectors/register', {
   'chelonia/private/queueEvent': function (queueName, invocation) {
     return sbp('okTurtles.eventQueue/queueEvent', queueName, ['chelonia/private/invoke', this._instance, invocation])
   },
-  'chelonia/private/loadManifest': async function (manifestHash: string) {
+  'chelonia/private/loadManifest': async function (contractName: string, manifestHash: string) {
+    if (!contractName || typeof contractName !== 'string') {
+      throw new Error('Invalid or missing contract name')
+    }
     if (this.manifestToContract[manifestHash]) {
       console.warn('[chelonia]: already loaded manifest', manifestHash)
       return
     }
     const manifestURL = `${this.config.connectionURL}/file/${manifestHash}`
-    const manifest = await fetch(manifestURL, { signal: this.abortController.signal }).then(handleFetchResult('json'))
+    const manifestSource = await fetch(manifestURL, { signal: this.abortController.signal }).then(handleFetchResult('text'))
+    const manifestHashOurs = createCID(manifestSource)
+    if (manifestHashOurs !== manifestHash) {
+      throw new Error(`bad hash ${manifestHashOurs} for manifst '${manifestHash}'! Should be: ${manifestHash}`)
+    }
+    const manifest = JSON.parse(manifestSource)
+    if (!has(manifest, 'signature') || typeof manifest.signature.keyId !== 'string' || typeof manifest.signature.value !== 'string') {
+      throw new Error(`Invalid or missing signature field for manifest ${manifestHash} (named ${contractName})`)
+    }
+    const rootState = sbp(this.config.stateSelector)
+    if (!has(rootState, 'contractSiginingKeys')) {
+      this.config.reactiveSet(rootState, 'contractSiginingKeys', Object.create(null))
+    }
+    const contractNameLookupKey = `name:${contractName}`
```

Comment:
Why not just use `contractName` directly?

Reply:
Because it's a user-defined key, so it's potentially insecure (e.g., if the name happens to be `__proto__`, `prototype` or something like that defined later.

Reply:
I don't think that's a problem. This key is coming from a trusted source (the developer's own source code), so it's fine.

If you can't trust the developer's code then it doesn't make sense to do anything at all, as there's no protection.